services/ui_backend_service/api/search.py

```diff
@@ -0,0 +1,268 @@
+import asyncio
+from typing import Tuple
+from services.data.db_utils import translate_run_key
+from services.data.tagging_utils import apply_run_tags_to_db_response
+from services.utils import handle_exceptions, logging
+from services.ui_backend_service.data.cache.utils import (
+    search_result_event_msg, error_event_msg
+)
+from .utils import query_param_enabled
+from urllib.parse import unquote_plus
+
+from aiohttp import web
+import json
+
+SCOPE_ARTIFACT = 'ARTIFACT'
+SCOPE_FOREACH_VARIABLE = 'FOREACH_VARIABLE'
+
+
+class SearchApi(object):
+    def __init__(self, app, db, cache=None):
+        self.db = db
+        app.router.add_route(
+            "GET", "/search/flows/{flow_id}/runs/{run_number}", self.get_run_tasks
+        )
+        self._artifact_table = self.db.artifact_table_postgres
+        self._metadata_table = self.db.metadata_table_postgres
+        self._run_table = self.db.run_table_postgres
+        self._artifact_store = getattr(cache, "artifact_cache", None)
+
+        self.search_results = list()
+        self.search_result_keyset = set()
+
+    @handle_exceptions
+    async def get_run_tasks(self, request):
+        self.search_results = list()
+        self.search_result_keyset = set()
```

Comment:
there's an issue with this. More comments in the `self.union_results`

Reply:
replied below.